geomesa-accumulo/geomesa-accumulo-datastore/src/main/scala/org/locationtech/geomesa/accumulo/iterators/ArrowBatchIterator.scala

```diff
@@ -44,78 +43,132 @@ class ArrowBatchIterator extends KryoLazyAggregatingIterator[ArrowBatchAggregate
+trait ArrowBatchAggregate {
+  def isEmpty: Boolean
+  def clear(): Unit
+  def add(sf: SimpleFeature): Unit
+  def size: Int
+  def encode(): Array[Byte]
+}
 
-  import org.locationtech.geomesa.arrow.allocator
+class ArrowBatchAggregateImpl(sft: SimpleFeatureType,
+                              dictionaries: Map[String, ArrowDictionary],
+                              encoding: SimpleFeatureEncoding) extends ArrowBatchAggregate {
 
-  import scala.collection.JavaConversions._
+  import org.locationtech.geomesa.arrow.allocator
 
   private var index = 0
 
-  private val vector = SimpleFeatureVector.create(sft, dictionaries, includeFids, GeometryPrecision.Float)
-  private val root = new VectorSchemaRoot(Seq(vector.underlying.getField), Seq(vector.underlying), 0)
-  private val unloader = new VectorUnloader(root)
-  private val os = new ByteArrayOutputStream()
+  private val vector = SimpleFeatureVector.create(sft, dictionaries, encoding)
+  private val batchWriter = new RecordBatchUnloader(vector)
 
-  def add(sf: SimpleFeature): Unit = {
+  override def add(sf: SimpleFeature): Unit = {
     vector.writer.set(index, sf)
     index += 1
   }
 
-  def isEmpty: Boolean = index == 0
+  override def isEmpty: Boolean = index == 0
 
-  def size: Int = index
+  override def size: Int = index
 
-  def clear(): Unit = {
-    vector.reset()
+  override def clear(): Unit = {
+    vector.clear()
     index = 0
   }
 
-  def encode(): Array[Byte] = {
-    os.reset()
-    vector.writer.setValueCount(index)
-    root.setRowCount(index)
-    MessageSerializer.serialize(new WriteChannel(Channels.newChannel(os)), unloader.getRecordBatch)
-    os.toByteArray
+  override def encode(): Array[Byte] = batchWriter.unload(index)
+}
+
+class ArrowSortingBatchAggregate(sft: SimpleFeatureType,
+                                 sortField: Int,
+                                 reverse: Boolean,
+                                 batchSize: Int,
+                                 dictionaries: Map[String, ArrowDictionary],
+                                 encoding: SimpleFeatureEncoding) extends ArrowBatchAggregate {
+
+  import org.locationtech.geomesa.arrow.allocator
+
+  private var index = 0
+  private val features = Array.ofDim[SimpleFeature](batchSize)
+
+  private val vector = SimpleFeatureVector.create(sft, dictionaries, encoding)
+  private val batchWriter = new RecordBatchUnloader(vector)
+
+  private val ordering = new Ordering[SimpleFeature] {
+    override def compare(x: SimpleFeature, y: SimpleFeature): Int = {
+      val left = x.getAttribute(sortField).asInstanceOf[Comparable[Any]]
+      val right = y.getAttribute(sortField).asInstanceOf[Comparable[Any]]
+      left.compareTo(right)
+    }
+  }
+
+  override def add(sf: SimpleFeature): Unit = {
+    // we have to copy since the feature might be re-used
+    // TODO we could probably optimize this...
+    features(index) = ScalaSimpleFeature.copy(sf)
```

Comment:
Could this approach just copy the backing byte arrays as an intermediate optimization?